controllers/bucket.go

```diff
@@ -0,0 +1,376 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllers
+
+import (
+	"errors"
+	"math"
+
+	hvpav1alpha1 "github.com/gardener/hvpa-controller/api/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	vpa_api "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1beta2"
+)
+
+const (
+	// ResourceCPU represents CPU in millicores (1core = 1000millicores).
+	ResourceCPU AxisName = "cpu"
+	// ResourceMemory represents memory, in bytes. (500Gi = 500GiB = 500 * 1024 * 1024 * 1024).
+	ResourceMemory AxisName = "memory"
+	// ResourceReplicas represents replicas.
+	ResourceReplicas AxisName = "replicas"
+)
+
+// Bucket defines the size and nature of bucket.
+// The argument "value" is product of values on "xAxis" and "yAxis".
+// A Bucket can have any number of axis, with their minValue and maxValue defined.
+// To find "XValue" for a given "value", we iterate over primary axis, the name for which should be passed as xAxis argument.
+// "YValue" would be the value in yAxis, such that XValue * YValue = "value".
+//
+// A Bucket is said to contain the "value" if:
+// 1. The "value" is greater than the product of minValues of xAxis and yAxis intervals, and
+// 2. The "value" is less than the product of maxValues of xAxis and yAxis intervals
+type Bucket interface {
```

Comment:
I would suggest to simplify this interface as well as the implementation to keep it closer to the problem domain (X axis is replicas and Y axis is vertical resource per replica) rather then a generic mathematical package. I think it would make both implementation and usage easier to understand.

I propose something like below.

```go
type EffectiveScalingIntervals interface {
  IsResourceInRangeForReplica(replicas int32, resourceName corev1.ResourceName, resourceValue int64) (bool, error)
  // If more than one replicas has the resource value in its range (due to overlap), then use biasReplicas
  // to resolve the conflict. If the biasReplicas is not among the candidate replicas then the lower of the replicas
  // should be returned.
  GetReplicasForResource(resourceName corev1.ResourceName, resourceValue int64, biasReplicas int32) (int32, error)
}
```

With the above interface, the factory method (I would suggest to call it `GetEffectiveScalingIntervals`) could return only `EffectiveScalingIntervals` and there is no need to return current bucket.

I would also suggest to simplify the implementation to use replicas as the X axis explicitly and call the Y axis resource. Also, as discussed out of band, please change the implementation to have one extrapolated effective scaling interval per replica value. Only the factory method should care about extrapolation types and not the rest of the implementation. This would help us easily test different parts of the implementation as well as easily introduce other types of extrapolation (exponential etc.) with minimal changes.

Reply:
done in https://github.com/gardener/hvpa-controller/pull/76/commits/6bb252bf07f9b57484d991ccfa66a8615e177408